File: rust/src/src/tool_security.rs
// src/tool_security.rs  <- src/tool_security.py
//! Server-side tool safety policy.

use once_cell::sync::Lazy;
use std::collections::HashSet;

// Tools regular/public users must not execute directly. These either expose
// server/runtime access, sensitive user data, external messaging, persistent
// state changes, or generic loopback/integration surfaces.
pub static NON_ADMIN_BLOCKED_TOOLS: Lazy<HashSet<&'static str>> = Lazy::new(|| {
    HashSet::from([
        "bash",
        "python",
        "read_file",
        "write_file",
        "search_chats",
        "manage_memory",
        "manage_skills",
        "manage_tasks",
        "manage_endpoints",
        "manage_mcp",
        "manage_webhooks",
        "manage_tokens",
        "manage_documents",
        "manage_settings",
        "api_call",
        "app_api",
        "send_email",
        "reply_to_email",
        "list_emails",
        "read_email",
        "resolve_contact",
        "manage_contact",
        "manage_calendar",
        "vault_search",
        "vault_get",
        "vault_unlock",
        "download_model",
        "serve_model",
        "serve_preset",
        "stop_served_model",
        "cancel_download",
        "adopt_served_model",
    ])
});

/// Return True when a non-admin/public user must not execute this tool.
pub fn is_public_blocked_tool(tool_name: Option<&str>) -> bool {
    let tool_name = match tool_name {
        Some(t) if !t.is_empty() => t,
        _ => return false,
    };
    NON_ADMIN_BLOCKED_TOOLS.contains(tool_name) || tool_name.starts_with("mcp__")
}

/// Return True for admins, or when auth is not configured yet.
pub fn owner_is_admin_or_single_user(owner: Option<&str>) -> bool {
    // The Python wraps the AuthManager lookup in try/except so that any failure
    // (import error, corrupt config, …) is logged via `logger.warning(...)` and
    // treated as "not admin". The Rust `AuthManager::new()` is infallible and the
    // body below cannot raise, so the `except` branch is unreachable here.
    let auth = crate::core::auth::AuthManager::new();
    if !auth.is_configured() {
        return true;
    }
    match owner {
        Some(o) if !o.is_empty() => auth.is_admin(o),
        _ => false,
    }
}

/// Tools to hide/disable for this owner under public-user policy.
pub fn blocked_tools_for_owner(owner: Option<&str>) -> HashSet<String> {
    if owner_is_admin_or_single_user(owner) {
        return HashSet::new();
    }
    NON_ADMIN_BLOCKED_TOOLS
        .iter()
        .map(|s| s.to_string())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serve_preset_is_blocked() {
        // A public/non-admin user must not be able to launch a model server.
        assert!(NON_ADMIN_BLOCKED_TOOLS.contains("serve_preset"));
        assert!(is_public_blocked_tool(Some("serve_preset")));
    }

    #[test]
    fn model_server_tools_are_all_blocked() {
        for tool in [
            "download_model",
            "serve_model",
            "serve_preset",
            "stop_served_model",
            "cancel_download",
            "adopt_served_model",
        ] {
            assert!(
                NON_ADMIN_BLOCKED_TOOLS.contains(tool),
                "expected {tool} to be in the non-admin blocklist",
            );
        }
    }
}

File: src/tool_security.py
"""Server-side tool safety policy."""

from __future__ import annotations

import logging
from typing import Optional, Set

logger = logging.getLogger(__name__)


# Tools regular/public users must not execute directly. These either expose
# server/runtime access, sensitive user data, external messaging, persistent
# state changes, or generic loopback/integration surfaces.
NON_ADMIN_BLOCKED_TOOLS = {
    "bash",
    "python",
    "read_file",
    "write_file",
    "edit_file",
    "grep",
    "glob",
    "ls",
    "search_chats",
    "manage_memory",
    "manage_skills",
    "manage_tasks",
    "manage_endpoints",
    "manage_mcp",
    "manage_webhooks",
    "manage_tokens",
    "manage_documents",
    "manage_settings",
    "api_call",
    "app_api",
    "send_email",
    "reply_to_email",
    "list_emails",
    "read_email",
    "resolve_contact",
    "manage_contact",
    "manage_calendar",
    "vault_search",
    "vault_get",
    "vault_unlock",
    "download_model",
    "serve_model",
    "serve_preset",
    "stop_served_model",
    "cancel_download",
    "adopt_served_model",
}


def context_has_untrusted(messages) -> bool:
    """True if any message in ``messages`` is untrusted-wrapped — content from
    outside the user's control: web pages, emails, RAG documents, research
    output, skill text, and the active editor document.

    Detection keys on the markers ``untrusted_context_message`` already sets:
    - the per-message ``metadata.trusted is False`` flag, and
    - the ``GUARD_OPEN`` delimiter literal embedded in the wrapped content (a
      backstop for paths where metadata is stripped before the model sees it).
    Because ``untrusted_context_message`` escapes that literal inside the body
    (see ``prompt_security._escape_guard_markers``), it only ever appears in a
    genuine wrapper — so the content check has no false positives.
    """
    from src.prompt_security import GUARD_OPEN

    for m in (messages or []):
        if not isinstance(m, dict):
            continue
        if (m.get("metadata") or {}).get("trusted") is False:
            return True
        content = m.get("content")
        if isinstance(content, str) and GUARD_OPEN in content:
            return True
    return False


def untrusted_attenuation_block(messages, *, enabled: bool) -> Set[str]:
    """Capability attenuation for prompt-injection defense (audit finding H2).

    When ``enabled`` AND the turn's context contains untrusted-wrapped content,
    drop the agent to public-user tool privileges — block
    ``NON_ADMIN_BLOCKED_TOOLS`` even for an admin / single-user owner — so a
    prompt injection that slips past the soft "untrusted" header still cannot
    reach high-impact tools (bash, vault, send_email, manage_tokens, model
    serving, ...). A hard gate at the tool layer, not a prompt instruction.

    Returns the empty set when disabled or when no untrusted content is present,
    so it is safe to leave the setting (agent_block_high_impact_on_untrusted)
    off by default — zero behaviour change until a user opts in.
    """
    if enabled and context_has_untrusted(messages):
        return set(NON_ADMIN_BLOCKED_TOOLS)
    return set()


# Plan mode: the agent may investigate but must not mutate anything. Only these
# read-only/inspection tools stay enabled; everything else (writes, sends,
# manage_*, model serving, MCP, etc.) is blocked. Allowlist rather than blocklist
# so any newly added tool defaults to BLOCKED in plan mode — fail safe.
#
# bash/python are deliberately NOT here: the shell can mutate (write files, hit
# the network) and can't be constrained to read-only at the tool layer, so plan
# mode blocks it outright rather than relying on a prompt to keep it well-behaved.
# Code/file discovery is covered by the dedicated read-only tools below
# (read_file, grep, glob, ls) instead of freestyle shell.
PLAN_MODE_READONLY_TOOLS = {
    "read_file",
    "grep",
    "glob",
    "ls",
    "web_search",
    "web_fetch",
    "search_chats",
    "list_models",
    "list_sessions",
    "list_emails",
    "read_email",
    "list_served_models",
    "list_downloads",
    "list_cached_models",
    "search_hf_models",
    "list_serve_presets",
    "list_cookbook_servers",
    "resolve_contact",
    "chat_with_model",
    "ask_teacher",
}


# The agent's tool gate is a DENYLIST: execute_tool_block blocks any tool whose
# name is in `disabled_tools`. Plan mode's policy is the opposite — an allowlist
# (PLAN_MODE_READONLY_TOOLS). To apply an allowlist through a denylist, plan mode
# returns the inverse: every known tool name minus the allowlist.
#
# Known tool names come from FUNCTION_TOOL_SCHEMAS, but that source is imperfect:
# some tools are only XML-invocable (e.g. manage_notes, generate_image) and never
# appear there, and the import can fail outright. Either gap would drop a mutating
# tool from the subtraction and silently leave it enabled. This set is the static
# backstop for both: union it in so known mutators are always subtracted, and so a
# failed import still blocks them (fail closed, never open). Only mutators belong
# here — read-only tools are covered by the allowlist. Keep in sync when adding
# new mutating tools.
_PLAN_MODE_KNOWN_MUTATORS = {
    "write_file", "create_document", "edit_document", "update_document",
    "suggest_document", "manage_documents", "create_session", "manage_session",
    "send_to_session", "pipeline", "manage_memory", "manage_skills",
    "manage_tasks", "manage_notes", "manage_endpoints", "manage_mcp",
    "manage_webhooks", "manage_tokens", "manage_settings", "manage_contact",
    "manage_calendar", "api_call", "app_api", "ui_control",
    "send_email", "reply_to_email", "bulk_email", "delete_email",
    "archive_email", "mark_email_read", "download_model", "serve_model",
    "stop_served_model", "cancel_download", "adopt_served_model", "serve_preset",
    "generate_image", "edit_image", "trigger_research", "manage_research",
    # Shell is never read-only-safe; block it explicitly so it stays out of plan
    # mode even if the schema list fails to load.
    "bash", "python",
}


def plan_mode_disabled_tools() -> Set[str]:
    """Tool names to add to the denylist in plan mode.

    Plan mode allows only PLAN_MODE_READONLY_TOOLS. The gate is a denylist, so
    return the inverse: every known tool name minus the allowlist. Known names
    come from the function-tool schemas, backstopped by _PLAN_MODE_KNOWN_MUTATORS
    (see above) so XML-only tools and a failed schema import can't leave a mutator
    enabled. MCP tools are handled separately — the loop drops the MCP manager
    entirely in plan mode."""
    try:
        # agent_tools / tool_parsing / tool_schemas form a mutually-circular
        # cluster that only resolves cleanly when entered via agent_tools.
        # Import it first so the lazy schema import works even from a cold
        # import (e.g. tests) — not just after the app has wired everything up.
        import src.agent_tools  # noqa: F401
        from src.tool_schemas import FUNCTION_TOOL_SCHEMAS

        all_names = {
            (t.get("function") or {}).get("name")
            for t in FUNCTION_TOOL_SCHEMAS
        }
        all_names.discard(None)
    except Exception as exc:
        logger.warning("Unable to load tool schemas for plan-mode gating: %s", exc)
        all_names = set()
    # Subtract the allowlist from all known tool names (schema-derived plus the
    # static mutator backstop). Fail closed: if the schema import failed above,
    # the backstop alone still blocks known mutators.
    return (all_names | _PLAN_MODE_KNOWN_MUTATORS) - PLAN_MODE_READONLY_TOOLS


def is_public_blocked_tool(tool_name: Optional[str]) -> bool:
    """Return True when a non-admin/public user must not execute this tool.

    This is a security gate, so it fails CLOSED: a malformed non-string tool
    name can't be matched against the blocklist or the ``mcp__`` namespace, so
    it is treated as blocked rather than silently allowed through. ``None`` /
    empty string means there is no tool to gate.
    """
    if tool_name is None or tool_name == "":
        return False
    if not isinstance(tool_name, str):
        return True
    return tool_name in NON_ADMIN_BLOCKED_TOOLS or tool_name.startswith("mcp__")


def owner_is_admin_or_single_user(owner: Optional[str]) -> bool:
    """Return True for admins, or when auth is not configured yet."""
    try:
        from core.auth import AuthManager

        auth = AuthManager()
        if not auth.is_configured:
            return True
        return bool(owner and auth.is_admin(owner))
    except Exception as exc:
        logger.warning("Unable to evaluate owner admin status: %s", exc)
        return False


def blocked_tools_for_owner(owner: Optional[str]) -> Set[str]:
    """Tools to hide/disable for this owner under public-user policy."""
    if owner_is_admin_or_single_user(owner):
        return set()
    return set(NON_ADMIN_BLOCKED_TOOLS)
